server/src/internalClusterTest/java/org/elasticsearch/snapshots/CloneSnapshotIT.java

```diff
@@ -469,6 +470,105 @@ public void testDoesNotStartOnBrokenSourceSnapshot() throws Exception {
         "] because its snapshot was not successful."));
     }
 
+    public void testStartSnapshotWithSuccessfulShardClonePendingFinalization() throws Exception {
+        final String masterName = internalCluster().startMasterOnlyNode(LARGE_SNAPSHOT_POOL_SETTINGS);
+        final String dataNode = internalCluster().startDataOnlyNode();
+        final String repoName = "test-repo";
+        createRepository(repoName, "mock");
+
+        final String indexName = "test-idx";
+        createIndexWithContent(indexName);
+
+        final String sourceSnapshot = "source-snapshot";
+        createFullSnapshot(repoName, sourceSnapshot);
+
+        blockMasterOnWriteIndexFile(repoName);
+        final String cloneName = "clone-blocked";
+        final ActionFuture<AcknowledgedResponse> blockedClone = startClone(repoName, sourceSnapshot, cloneName, indexName);
+        waitForBlock(masterName, repoName, TimeValue.timeValueSeconds(30L));
+        awaitNumberOfSnapshotsInProgress(1);
+        blockNodeOnAnyFiles(repoName, dataNode);
+        final ActionFuture<CreateSnapshotResponse> otherSnapshot = startFullSnapshot(repoName, "other-snapshot");
+        awaitNumberOfSnapshotsInProgress(2);
+        assertFalse(blockedClone.isDone());
+        unblockNode(repoName, masterName);
+        awaitNumberOfSnapshotsInProgress(1);
+        awaitMasterFinishRepoOperations();
+        unblockNode(repoName, dataNode);
+        assertAcked(blockedClone.get());
+        assertEquals(getSnapshot(repoName, cloneName).state(), SnapshotState.SUCCESS);
+        assertSuccessful(otherSnapshot);
```

Comment:
I guess that it's enough to assert that the snapshot is successful to ensure that they get the corresponding generation number? otherwise it would be a partial snapshot, right?

Reply:
Yea exactly, if the snapshot was based off of a broken/outdated shard generation value it would be `PARTIAL` because the shard(s) working from the broken generation would fail with a ~`FileNotFoundException`/404.